README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# penngradlings - Themes and utilities for Penn GradLingS

<!-- badges: start -->

[![R-CMD-check](https://github.com/yjunechoe/penngradlings/workflows/R-CMD-check/badge.svg)](https://github.com/yjunechoe/penngradlings/actions)
<!-- badges: end -->

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("yjunechoe/penngradlings")
```

## Example

``` r
library(penngradlings)
#> Package fonts loaded. See `?load_pkg_fonts` for details.
```

``` r
library(ggplot2)

ggplot(diamonds, aes(cut, price, fill = cut)) +
  stat_summary(geom = "bar", fun = "mean", width = 0.8) +
  scale_y_continuous(
    expand = expansion(c(0, 0.1)),
    labels = ~ paste0("$", .x)
  ) +
  scale_fill_pgl_discrete(what = "cat_coaster", who = "xin") +
  labs(
    title = "Price of diamonds by cut quality",
    subtitle = "This is a bad plot because carat (size) is a confounding variable",
    x = NULL, y = NULL,
    caption = "diamonds dataset from {ggplot2}"
  ) +
  theme_pgl_minimal(axis_lines = "x", grid_lines = "y")
```

<img src="man/figures/README-theme-minimal-1.png" width="100%" />

More on the [package
website](https://yjunechoe.github.io/penngradlings).


NEWS.md
# penngradlings (development version)

# penngradlings 0.2.2

### New functions

- `fct_derive()` to derive levels from another factor with unique pair-wise matches

### Enhancements

- `read_pcibex()` gains an `encoding` argument for specifying the file's character encoding (#18)
- `theme_pgl_minimal()` has been re-written to scale better, default value for `base_size` changed to 12

# penngradlings 0.2.1

### New functions

- `raggsave_auto()` to save any arbitrary plot with `ragg::agg_png()`

### Enhancements

- `read_pcibex()` gains an `encoding` argument for specifying the file's character encoding. 
- `ggsave_auto()` supports auto rendering of grobs

# penngradlings 0.2.0

### New functions

- `scale_color/fill_pgl_discrete()`

### Enhancements

- Overhaul of `pgl_pals()`. It's now a function factory

# penngradlings 0.1.1

### New functions

- `name_self()` list-constructor function, mostly to help with naming multiple columns when passing a list of functions to `dplyr::across()`.

- `read_pcibex()` for reading in the results from a PCIbex experiment. 

### Enhancements

- You can now set global options for `ggsave_auto()` with `options(ggsave_auto.*)`. The package now also comes with an RStudio addin for calling `ggsave_auto()` (one possible keyboard binding is Ctrl/Cmd+Shift+G, which overwrites the "Find Previous" shortcut).

- `theme_pgl_minimal()` gets `strip.placement = outside` to hack margins into the space between facet and panel with a blank secondary axis.

# penngradlings 0.1.0

First release.

### New functions

#### Fonts

- The package packs 8 open source font families: _Atkinson Hyperlegible_, _Bootstrap Icons_, _Charis SIL_, _Font Awesome 5 Free_, _Inter_, _Material Icons_, _Piazzolla_, _Piazzolla SC_. Fonts are automatically registered on package load. See `.pgl_fonts()` for details.
- A `font_*()` family of helper functions for inspecting, registering, and shaping fonts - `font_hoist()`, `font_pluck()`, `font_info()`, `font_exists()`, and `font_path()`.

#### Themes

- `theme_pgl_basic()`, a low-ink theme design built on top of `ggplot2::theme_minimal()`, with options for turning on/off axis and grid lines.

#### Layers

- Stylized panel annotations with `geom_text_outline()` and `annotation_textbox()`.

#### Save/render

- `ggsave_auto()`, a wrapper around `ggplot2::ggsave()` offering convenience for interactive plotting. Accumulates all plots rendered with the function in a separate _./ggsave_auto_ directory and opens the saved figure back using the system's default photo-viewing app. Includes option for reading the saved figure back as a {magick} object.


R/ggplot2-themes.R
#' Basic {penngradlings} theme template
#'
#' @param base_size Base size for text elements
#' @param axis_lines Which axis lines should be drawn, passed as a string.
#' @param grid_lines Which grid lines should be drawn, passed as a string.
#' @param grid_lines_minor Whether minor grid lines should be draw. Defaults to `FALSE`
#' @param outline_color Primary color for text, line, and rectangle outline.
#'
#' @return A theme object created by `theme()`
#'
#' @importFrom ggplot2 theme element_text element_line element_rect element_blank margin rel
#' @export
#'
#' @examples
#' \dontrun{
#' library(ggplot2)
#'
#' ggplot(iris, aes(Petal.Length, Petal.Width, color = Species)) +
#'   geom_point(size = 2) +
#'   labs(
#'     title = "Petal profile of plant species in the iris dataset",
#'     subtitle = "Virginica has the largest petal of the three species"
#'   ) +
#'   facet_wrap(~Species) +
#'   theme_pgl_minimal(base_size = 11)
#'
#' ggplot(diamonds, aes(cut, price, fill = cut)) +
#'   stat_summary(geom = "bar", fun = "mean", width = 0.8) +
#'   scale_y_continuous(
#'     expand = expansion(c(0, 0.1)),
#'     labels = ~ paste0("$", .x)
#'   ) +
#'   scale_fill_manual(values = pgl_pals("christine", "blueberry_boba")) +
#'   labs(
#'     title = "Price of diamonds by cut quality",
#'     subtitle = "This is a bad explanatory plot because carat (size) is a confounding variable",
#'     x = NULL, y = NULL,
#'     caption = "diamonds dataset from {ggplot2}"
#'   ) +
#'   theme_pgl_minimal(axis_lines = "x", grid_lines = "y")
#' }
theme_pgl_minimal <- function(base_size = 12,
                              axis_lines = "",
                              grid_lines = "xy",
                              grid_lines_minor = FALSE,
                              outline_color = "#1F1C1C") {

  faded_outline_color <- colorspace::lighten(outline_color, 0.35)
  faded_text_color <- colorspace::lighten(outline_color, 0.15)

  pad_in_sm <- base_size/200
  pad_in_md <- base_size/100
  pad_in_lg <- base_size/50

  .theme <- ggplot2::theme_minimal(
    base_size = base_size,
    base_family = "Inter-Medium"
  ) +
    theme(
      rect = element_rect(color = faded_outline_color),
      line = element_line(color = faded_outline_color),
      text = element_text(color = outline_color, family = "Inter-Medium"),
      title = element_text(color = faded_text_color, family = "Inter-Bold"),
      plot.title = element_text(size = rel(1.2), family = "Inter-ExtraBold", margin = margin(b = pad_in_md, unit = "in")),
      plot.background = element_rect(color = NA, fill = "white"),
      plot.subtitle = element_text(color = faded_outline_color, margin = margin(t = 0, b = pad_in_lg, unit = "in")),
      plot.caption = element_text(color = faded_text_color, family = "Inter-Regular", margin = margin(t = pad_in_md, unit = "in")),
      plot.title.position = "plot",
      plot.caption.position = "plot",
      plot.tag = element_text(color = outline_color, size = rel(1.5), family = "Inter-Bold", margin = margin(b = pad_in_sm, r = pad_in_sm, unit = "in")),
      plot.margin = margin(pad_in_md, pad_in_md, pad_in_md, pad_in_md, "in"),
      axis.text = element_text(family = "Inter-SemiBold"),
      axis.text.x = element_text(size = rel(1)),
      axis.text.y = element_text(size = rel(0.9)),
      axis.text.x.bottom = element_text(margin = margin(t = pad_in_sm, unit = "in")),
      axis.text.x.top = element_text(margin = margin(b = pad_in_sm, unit = "in")),
      axis.text.y.left = element_text(margin = margin(r = pad_in_sm, unit = "in")),
      axis.text.y.right = element_text(margin = margin(l = pad_in_sm, unit = "in")),
      axis.title.x.bottom = element_text(margin = margin(t = pad_in_sm, unit = "in")),
      axis.title.x.top = element_text(margin = margin(b = pad_in_sm, unit = "in")),
      axis.title.y.left = element_text(margin = margin(r = pad_in_sm, unit = "in")),
      axis.title.y.right = element_text(margin = margin(l = pad_in_sm, unit = "in")),
      panel.grid = element_line(color = "#F2F2F2"),
      panel.spacing = grid::unit(pad_in_sm, "in"),
      legend.title = element_text(family = "Inter-SemiBold"),
      legend.text = element_text(color = faded_text_color, size = rel(0.8), family = "Inter-SemiBold"),
      legend.key = element_blank(),
      strip.placement = "outside",
      strip.background = element_rect(),
      strip.text = element_text(size = rel(0.85), family = "Inter-SemiBold")
    )

  if (nchar(axis_lines) <= 2 && grepl("[xyXY]{1,2}", axis_lines)) {
    if (grepl("(x|X)", axis_lines)) {
      .theme <- .theme +
        theme(axis.line.x = element_line())
    }
    if (grepl("(y|Y)", axis_lines)) {
      .theme <- .theme +
        theme(axis.line.y = element_line())
    }
  }

  if (nchar(grid_lines) <= 2 && grepl("[xyXY]{0,2}", grid_lines)) {
    if (!grepl("(x|X)", grid_lines)) {
      .theme <- .theme +
        theme(
          panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank()
        )
    }
    if (!grepl("(y|Y)", grid_lines)) {
      .theme <- .theme +
        theme(
          panel.grid.major.y = element_blank(),
          panel.grid.minor.y = element_blank()
        )
    }
  }

  if (!grid_lines_minor) {
    .theme <- .theme +
      theme(panel.grid.minor = element_blank())
  }

  .theme
}
